revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,293 @@
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
+}
+
+func certCheckStatus(cert, issuer *x509.Certificate, opts Options) *base.CertRevocationResult {
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		return &base.CertRevocationResult{
+			Result:        base.NonRevokable,
+			ServerResults: []*base.ServerResult{errToServerResult(NoOCSPServerError{})},
+		}
+	}
+
+	serverResults := make([]*base.ServerResult, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		serverResult := checkStatusFromServer(cert, issuer, server, opts)
+		switch t := serverResult.Error.(type) {
+		case nil, RevokedError, UnknownStatusError:
+			// A valid response has been received from an OCSP server
+			// Result should be based on only this response, not any errors from
+			// other servers
+			return serverResultsToCertRevocationResult([]*base.ServerResult{errToServerResult(t)})
+		default:
+			serverResults[serverIndex] = serverResult
+			continue
+		}
+
+	}
+	return serverResultsToCertRevocationResult(serverResults)
+}
+
+func checkStatusFromServer(cert, issuer *x509.Certificate, server string, opts Options) *base.ServerResult {
+	// Check valid server
+	if serverURL, err := url.Parse(server); err != nil || strings.ToLower(serverURL.Scheme) != "http" {
+		// This function is only able to check servers that are accessible via HTTP
+		return errToServerResult(OCSPCheckError{Err: fmt.Errorf("OCSPServer protocol %s is not supported", serverURL.Scheme)})
+	}
+
+	// Create OCSP Request
+	resp, err := executeOCSPCheck(cert, issuer, server, opts)
+	if err != nil {
+		// If there is a server error, attempt all servers before determining what to return
+		// to the user
+		return errToServerResult(err)
+	}
+
+	// Validate OCSP response isn't expired
+	if time.Now().After(resp.NextUpdate) {
+		return errToServerResult(OCSPCheckError{Err: errors.New("expired OCSP response")})
+	}
+
+	// Check for presence of pkix-ocsp-no-check extension in response
+	foundNoCheck := false
+	for _, extension := range resp.Extensions {
+		if !foundNoCheck && extension.Id.String() == pkixNoCheckOID {
+			foundNoCheck = true
+			break
+		}
+	}
+	if !foundNoCheck {
+		// This will be ignored until CRL is implemented
+		// If it isn't found, CRL should be used to verify the OCSP response
+		// TODO: add CRL support
+		// https://github.com/notaryproject/notation-core-go/issues/125
+		return errToServerResult(PKIXNoCheckError{})
+	}
+
+	// Check if SigningTime was before the revocation
+	if opts.SigningTime.Before(resp.RevokedAt) {
+		return errToServerResult(nil)
+	}
+
+	// No errors, valid server response
+	switch resp.Status {
+	case ocsp.Good:
+		return errToServerResult(nil)
+	case ocsp.Revoked:
+		return errToServerResult(RevokedError{})
```

Comment:
I might be missing it, but in our spec:
https://github.com/notaryproject/notaryproject/blob/main/specs/trust-store-trust-policy.md#ocsp-download, it says `If the certificate is revoked i.e CertStatus is revoked, look for InvalidityDate. If the invalidity date is present and timestamp signature is also present then if the invalidity date is before the timestamping date, the certificate is considered revoked. If the invalidity date is not present in OCSP response, the certificate is considered revoked.`
I couldn't find this part in the PR. Are we implementing this somewhere?

Reply:
This check is actually performed right above this switch statement on line 170. It looks at the RevokedAt field for the response, which is how the InvalidityDate is expressed in an ocsp.Response object

Reply:
> After the newest commits, it is now on line 149. This is the check being done: `if opts.SigningTime.Before(resp.RevokedAt) {`

Okay, if I understand this correctly, your implementation corresponds to the following:
```
If the certificate is revoked i.e CertStatus is revoked, look for InvalidityDate. If the invalidity date is present and timestamp signature is also present then if the invalidity date is AFTER the timestamping date, the certificate is considered NOT revoked. If the invalidity date is not present in OCSP response, the certificate is considered revoked.
```
which is a bit different from the spec above? (maybe we should rephrase the spec)
In addition, I notice that if `opts.SigningTime.Before(resp.RevokedAt)` returns true, we are not checking `resp.Status` anymore. Should we always check `resp.Status` and move this logic under `ocsp.Revoked` instead? So the code becomes:
```
// No errors, valid server response
switch resp.Status {
case ocsp.Good:
  return errToServerResult(nil)
case ocsp.Revoked:
  // Check if SigningTime was before the revocation
  if opts.SigningTime.Before(resp.RevokedAt) {
    return errToServerResult(nil)
  }
  return errToServerResult(RevokedError{})
default:
  // ocsp.Unknown
  return errToServerResult(UnknownStatusError{})
}
```

Reply:
I agree. I think this was a carry over from CRL, which is noted in RFC 5280 "The invalidity date is a non-critical CRL entry extension that provides the date on which it is known or suspected that the private key was compromised or that the certificate otherwise became invalid", but OCSP uses revocationTime per RFC6960 "MUST specify the revocationTime" (this is referenced in the x509.Certificate structure as RevokedAt). So yes, we should update the spec to reflect this difference. //cc: @priteshbandi 

W.r.t moving the time check inside the status check, I think that makes sense. I'll add it to the next revision

Reply:
Same here, we need to open an issue to rephrase our spec in https://github.com/notaryproject/notaryproject. @kody-kimberl

Reply:
Update: New commit aligns the code with the wording of the spec. Potential updates to spec are being tracked here: https://github.com/notaryproject/notation-core-go/pull/134#discussion_r1169389012. Resolving